vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -1,23 +1,153 @@
 #include <algorithm>
 
+using namespace LEARNER;
 using namespace VW;
+using namespace VW::config;
+
+template <bool is_learn>
+void do_actual_learning(CCB::ccb& data, multi_learner& base, multi_ex& examples) {
+
+  //get decisions, actions and shared parts of the multiline example
+  std::vector<example*> decisions, actions;
+  example* shared;
+  for (auto ex : examples) {
+    switch (ex->l.conditional_contextual_bandit.type) {
+      case CCB::example_type::shared:
+        shared = ex;
+        break;
+      case CCB::example_type::action:
+        actions.push_back(ex);
+        break;
+      case CCB::example_type::decision:
+        decisions.push_back(ex);
+        break;
+      default:
+        THROW("ccb_adf_explore: badly formatted example - invalid example type");
+    }
+  }
 
-bool CCB::ec_is_example_header(example& ec)
-{
-  return ec.l.conditional_contextual_bandit.type == example_type::shared;
+  //sanity check
+  if (decisions.size() > actions.size())
+    THROW("ccb_adf_explore: badly formatted example - number of actions " << actions.size() << " must be greater than the number of decisions " << decisions.size());
+
+  //TODO inject decisions into shared (share := shared + decision features)
+  //TODO add tests and UT
+
+  //set the CB label in the shared example
+  //set the proba to -1 and the cost to float_max (convention for CB to identify shared data)
+  CB::cb_class f;
+  f.probability = -1.f;
+  f.cost = FLT_MAX;
+  shared->l.cb.costs = v_init<CB::cb_class>();
+  shared->l.cb.costs.push_back(f);
+
+  std::set<uint32_t> blacklist_actions, whitelist_actions;
+  CCB::decision_scores_t decision_scores = v_init< ACTION_SCORE::action_scores>();
+
+  //for each decision, attach the label to the chosen action and perform a CB call
+  for (auto decision : decisions) {
+    multi_ex cb_examples;//multiline example for the cb_explore_adf call
+    cb_examples.push_back(shared);
+
+    //sanity check
+    if (is_learn) {
+      if (decision->l.conditional_contextual_bandit.outcome == nullptr)
+        THROW("ccb_adf_explore: badly formatted example - missing label");
+      if (decision->l.conditional_contextual_bandit.outcome->probabilities.size() == 0)
+        THROW("ccb_adf_explore: badly formatted example - missing label probability");
+    }
+
+    //read the ccb label
+    if (is_learn) {
+      f.cost = decision->l.conditional_contextual_bandit.outcome->cost;
+      f.probability = decision->l.conditional_contextual_bandit.outcome->probabilities[0].score;
+      f.action = decision->l.conditional_contextual_bandit.outcome->probabilities[0].action;
+    }
+    whitelist_actions.clear();
+    for (uint32_t a : decision->l.conditional_contextual_bandit.explicit_included_actions) whitelist_actions.insert(a);
+
+    for (auto action : actions) {
+      action->l.cb.costs = v_init<CB::cb_class>();//set a default (valid) cb label
+      if (is_learn && f.action == action->example_counter) //set the label for the chosen action
```

Comment:
The semantics of action->example_counter are what?   Normally example_counter is something that increments over _all_ examples.

Reply:
Yes, example_counter is incremented over all example in the multi-example.
For example with the input:
```
ccb shared | s
ccb action | a1
ccb action | a2
ccb action | a3
ccb decision 2:-1:0.5 | d1
ccb decision 1:0:0.5 | d2
```
The first line will have example_counter = 0, and the last line example_counter = 5.

Reply:
There are two issues here.
1) example_counter is a bad name for an action.
2) this field is set by the system, so it's meaning can change according to the system.  As such, we shouldn't rely on it.